artcommon/artcommonlib/konflux/konflux_db.py

```diff
@@ -36,6 +321,107 @@ def bind(self, record_cls: typing.Type[KonfluxRecord]):
 
+    async def _ensure_group_cached(self, group: str):
+        """
+        Lazy-load cache for group if not already loaded.
+
+        Automatically loads the last N days of builds for the group on first access.
+        Uses per-group async events to coordinate concurrent load attempts - only one
+        coroutine will actually load while others wait for it to complete.
+
+        :param group: Group name (e.g., 'openshift-4.18')
+        """
+        if not self.cache:
+            return
+
+        # Quick check - if already loaded, return immediately
+        if self.cache.is_group_loaded(group):
+            self.logger.debug(f"Cache already loaded for group '{group}'")
+            return
+
+        # Get or create an event for this group (thread-safe)
+        with KonfluxDb._cache_lock:
+            if group not in KonfluxDb._group_loading_events:
+                # Create new event for this group
+                event = asyncio.Event()
+                KonfluxDb._group_loading_events[group] = event
+                should_load = True
+            else:
+                # Another coroutine is already loading
+                event = KonfluxDb._group_loading_events[group]
+                should_load = False
+
+        if not should_load:
+            # Wait for the other coroutine to finish loading
+            self.logger.debug(f"Waiting for another coroutine to load group '{group}'...")
+            await event.wait()
+            self.logger.debug(f"Cache loaded for group '{group}' by another coroutine")
+            return
+
+        # We're the one who will load
+        try:
+            self.logger.info(f"Lazy-loading cache for group '{group}' (last {self.cache._cache_days} days)...")
+
+            # Build query for last N days of builds in this group
+            start_time = datetime.now(tz=timezone.utc) - timedelta(days=self.cache._cache_days)
+            where_clauses = [
+                Column('outcome', String).in_(['success', 'failure']),
+                Column('group', String) == group,
+                Column('start_time', DateTime) >= start_time,
+            ]
+
+            order_by_clause = Column('start_time', quote=True).desc()
+
+            # Execute single large query
+            rows = await self.bq_client.select(
+                where_clauses=where_clauses,
+                order_by_clause=order_by_clause,
+                limit=None,  # Get all results
+            )
+
+            # Load all rows into cache (thread-safe operation)
+            builds = [self.from_result_row(row) for row in rows]
+            self.cache.add_builds(builds, group)
+
+            self.logger.info(f"Cache loaded for group '{group}': {len(builds)} builds")
+
+        except Exception as e:
+            self.logger.error(f"Failed to load cache for group '{group}': {e}")
+            raise
+        finally:
+            # Signal completion to any waiting coroutines
+            event.set()
```

Comment:
Here the event is set no matter whether loading cache is a success. If the current coroutine fails to load, other coroutines will always go to the `if not should_load:` branch and prevent from loading again. 

Some improvements could be:
1. Allow another coroutine to reload if the current coroutine fails to load.
2. The event can be removed from _group_loading_events after the group is cached. Other coroutines can recheck whether loading is needed with is_group_loaded inside of _cache_lock after the quick check at line 337.

Reply:
Updated. Thanks!